### PR TITLE
Feature/dock 1765/entry page category buttons

### PR DESCRIPTION
**Pull request by svonworl**
**Description**
Third and final Categories PR of the initial UI implementation, which adds category buttons to the workflow and tool pages.  

The appearance is pretty close to the mockups, except that the indentation of the button row is similar to the existing "Last Modified" text directly above, so it matches.   It looked funny otherwise.

**Update: the below has been edited to reflect a new approach @ 2021/Nov/19 3:35pm PST**

To avoid a race condition bug, `EntryCategoriesService.updateEntryCategories()` is implemented slightly differently than other similar existing methods.  Before the HTTP request, the current update subscription (if it exists) is unsubscribed, and then set to the subscription for the new HTTP request.  This avoids the case where the updates for entry A and B overlap, the responses get reordered, and the categories of A are stored last and displayed on B's page.  Depending on the protocols and software involved, this could happen due to network problems and/or thread scheduling vagaries.  Imagine the user quickly navigating to one entry page and then another, and a server thread receiving the first categories request, and then blocking or not being scheduled again until another thread completes the second categories request.  Or, suppose the user is connected via HTTP/3-QUIC and packet loss delays the first request/response.  Similarly, HTTP/1.1 with multiple connections.  This kind of thing doesn't happen much, but it does happen.

This type of bug probably affects most of the existing Service methods that retrieve and store collections of objects.  

**Issue**
dockstore/dockstore#4476
https://ucsc-cgl.atlassian.net/browse/DOCK-1765

**Screenshots**
Screen Shot 2021-11-18 at 5 35 01 PM

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that your code compiles by running `npm run build`
- [x] If this is the first time you're submitting a PR or even if you just need a refresher, consider reviewing our [style guide](https://github.com/dockstore/dockstore/wiki/Dockstore-Frontend-Opinionated-Style-Guide#pr-checklist)
- [x] Do not bypass Angular sanitization (bypassSecurityTrustHtml, etc.), or justify why you need to do so
- [x] If displaying markdown, use the `markdown-wrapper` component, which does extra sanitization
- [x] Do not use cookies, although this may change in the future
- [ ] Run `npm audit` and ensure you are not introducing new vulnerabilities
- [x] Do due diligence on new 3rd party libraries, checking for CVEs
- [x] Don't allow user-uploaded images to be served from the Dockstore domain

### CODE REVIEW

**Review by garyluu — COMMENTED** [review]
if the edge case is going to be something you want to commonly handle, make a test for it.

additionally, it looks like if it does happen, you're stuck in an infinite loading state?

**Comment by svonworl** [comment]
I don't think it results in an infinite loading state - the loading state should be set by the handlers for the response to the most recent request.  However, I realized...

A simpler solution is to cancel the previous subscription before each successive call.  Same effect, easier to understand.  I will implement it, push, and update the PR description.

**Comment by garyluu** [comment]
> I don't think it results in an infinite loading state - the loading state should be set by the handlers for the response to the most recent request. However, I realized...
> 
> A simpler solution is to unsubscribe the previous subscription before each successive call. Same effect, easier to understand. I will implement it, push, and update the PR description.

I misread a statement in the code. shouldn't have an infinite loading state.

**Review by coverbeck — COMMENTED** [review]
Was here, going to wait for

> A simpler solution is to unsubscribe the previous subscription before each successive call. Same effect, easier to understand. I will implement it, push, and update the PR description.

You may mean this, but I think the `switchMap` suggested by Gary would accomplish this.

More generally, maybe I'm wrong, but I feel this is pretty edge-casey and sometimes the nature of the web. What if you have two tabs open; the first tab won't reflect changes made in the second tab, etc.

Unless we move to web sockets, I don't think there is an all-encompassing solution, and I wouldn't  recommend spending  a lot of time at this point trying to figure this out.

But if your pattern works, isn't too hard to do, and can easily be applied elsewhere, I don't object.

**Comment by svonworl** [comment]
I do agree it's an edge case, but when it happens, it results in the wrong information being displayed.  Doesn't matter much in this case (the wrong categories get displayed), but if something more important got foobared, it could be a bigger issue.  Maybe it would cause the user to accidentally modify or delete something, or trigger a bug that we spend days to track down, ...

In any case, the revised fix is pretty simple, and understandable enough so that it might not even need a test.  Take a look and let me know what you think.

File: src/app/container/container.component.html [hunk 1]
```diff
@@ -104,9 +104,14 @@ <h3>
     </div>
   </div>
   <div class="row m-1">
-    <div *ngIf="tool" class="col-sm-12" style="margin-bottom: 5px">
-      <span *ngIf="!labelsEditMode && !starGazersClicked">
-        <mat-chip-list>
+    <div *ngIf="tool">
+      <span *ngIf="(categories$ | async).length > 0" class="inline-block mr-5 mt-1 mb-1">
+        <span class="size-12 mr-2">Categories</span>
+        <app-category-button *ngFor="let category of (categories$ | async)" [category]="category" entryType="tool" class="weight-500"></app-category-button>
```
Comment (garyluu) [inline src/app/container/container.component.html:110]: aren't all app-category-button instances visually identical? How come this has a different font-weight?

File: src/styles.scss [hunk 2]
```diff
@@ -851,6 +851,20 @@ body {
   vertical-align: middle;
 }
 
+// Set various font attributes
+.weight-500 {
+  font-weight: 500;
+}
+
+.size-12 {
```
Comment (garyluu) [inline src/styles.scss:859]: PR checklist, need a comment for this.

From what I can tell, the mockups (not including headings) only use 14px (default) or 12px. There's no 13, 11.9, etc. If true and in order to maintain consistency, the comment should mention the use case for this.  Would hopefully prevent others from using `<small`, or font-size: small, font-size: smaller, and we would use this else where instead.

**Review by garyluu — COMMENTED** [review]
new solution suprisingly works as well as switchmap as far as i can tell and is simpler unlike the first solution (doesn't cancel http responses)

File: src/app/categories/state/entry-categories.service.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  private currentSubscription: Subscription = null;
+
+  /**
+   * Updates the list of categories for the specified entry.
+   */
+  updateEntryCategories(entryId: number) {
+    this.entryCategoriesStore.setLoading(true);
+    this.entryCategoriesStore.setError(false);
+    this.entryCategoriesStore.remove();
+    this.currentSubscription?.unsubscribe();
```
Comment (coverbeck) [inline src/app/categories/state/entry-categories.service.ts:49]: If `updateEntryCategories` were invoked with two different entry ids in succession, the response for the first entry could get unsubscribed prematurely, and its response ignored. But it looks like the store only has categories for one entry anyway, it may not matter and might actually be the desired behavior.

**Review by coverbeck — APPROVED** [review]
Pending addressing of Gary's CSS comments.

File: src/app/categories/state/entry-categories.service.ts (hunk 3 above)
Comment (svonworl) [reply src/app/categories/state/entry-categories.service.ts:49]: What you describe is the desired behavior.  I implemented this service/store similarly to the existing services/stores, which store a single unit of information which ostensibly corresponds to an item on the current page.  Essentially, my little tweak ignores the first response if it arrives after another request is launched, because we've changed/updated the page and don't want/need the first response any more.

I have some ideas about dynamically-allocated stores which don't suffer from this race condition, and would allow us to do things like retrieve the categories for each of multiple entries on the same page, but that's a discussion for some other day...

File: src/app/container/container.component.html (hunk 1 above)
Comment (svonworl) [reply src/app/container/container.component.html:110]: Currently, the app-category-buttons inherit the font-weight, which was a choice I made after misinterpreting the mockups, and probably a mistake.  I am going to harmonize them all to the same weight, and eliminate the weight-500 style.  Will flag you for re-review after I push.

**Review by garyluu — APPROVED** [review]
